lib/bamboo/template.ex

```diff
@@ -0,0 +1,271 @@
+defmodule Bamboo.Template do
+  @moduledoc """
+  Render emails with layouts, view modules, and templates.
+
+  This module allows rendering emails with layouts and views. Pass an
+  atom (e.g. `:welcome_email`) as the template name to render both HTML and
+  plain text emails. Use a string if you only want to render one type, e.g.
+  `"welcome_email.text"` or `"welcome_email.html"`.
+
+  ## Examples
+
+  _Set the text and HTML layout for an email._
+
+      defmodule MyApp.Email do
+        use Bamboo.Template, view: MyApp.Email.AccountView
+
+        def welcome_email do
+          new_email()
+          |> put_text_layout({MyApp.Email.LayoutView, "email.text"})
+          |> put_html_layout({MyApp.Email.LayoutView, "email.html"})
+          |> render(:welcome) # Pass atom to render html AND plain text templates
+        end
+      end
+
+  _Set both the text and HTML layout at the same time for an email._
+
+      defmodule MyApp.Email do
+        use Bamboo.Template, view: MyApp.Email.AccountView
+
+        def welcome_email do
+          new_email()
+          |> put_layout({MyApp.Email.LayoutView, :email})
+          |> render(:welcome)
+        end
+      end
+
+  _Render both text and html emails without layouts._
+
+      defmodule MyApp.Email do
+        use Bamboo.Template, view: MyApp.Email.AccountView
+
+        def welcome_email do
+          new_email()
+          |> render(:welcome)
+        end
+      end
+
+  _Make assigns available to a template._
+
+      defmodule MyApp.Email do
+        use Bamboo.Template, view: MyApp.Email.AccountView
+
+        def welcome_email(user) do
+          new_email()
+          |> assign(:user, user)
+          |> render(:welcome)
+        end
+      end
+
+  _Make assigns available to a template during render call._
+
+      defmodule MyApp.Email do
+        use Bamboo.Template, view: MyApp.Email.AccountView
+
+        def welcome_email(user) do
+          new_email()
+          |> put_html_layout({MyApp.Email.LayoutView, "email.html"})
+          |> render(:welcome, user: user)
+        end
+      end
+
+  _Render an email by passing the template string to render._
+
+      defmodule MyApp.Email do
+        use Bamboo.Template, view: MyApp.Email.AccountView
+
+        def html_email do
+          new_email()
+          |> render("html_email.html")
+        end
+
+        def text_email do
+          new_email
+          |> render("text_email.text")
+        end
+      end
+
+  ## HTML Layout Example
+
+      # lib/my_app/email.ex
```

Comment:
Nitpick: I would expect to find email-related modules is `lib/my_app_web/**/*`. That presumes the Phoenix convention to some extent. But I'm wondering if listing the file paths should either follow the most dominant convention (i.e. Phoenix) or should be left out. Thoughts?

Reply:
I wanted to keep away from any notion of `my_app_web` here because the whole purpose of this feature is for people to use this who _don't_ use Phoenix. Emailing is not always a web concern for people. We use HTML to render some of those emails (though it could also just be text), but not all apps have a "web" layer. What do you think? Still think we should include `my_app_web`?